pytestCheckHook: support inclusion and exclusion of path globs, test items, keywords, and markers

ShamrockLee opened the pull request:
If applied, pytestCheckHook will accept
- `enabledTestPaths` and `disabledTestPaths` to include/exclude either path globs or test items (`<file path>::<specifier>`)
- `enabledTests` and `disabledTests` to include/exclude keywords
- `enabledTestMarks` and `disabledTestMarks` to include/exclude markers

The inclusion corresponds to the positional arguments of `pytest`, as described by [the PyTest documentation](https://docs.pytest.org/en/stable/how-to/usage.html).

The path globs inside `enabledTestPaths` is expanded using an in-line Python script powered by the `glob` module from the Python Standard Library. `enabledTestPaths` is crucial for deprecating `pytestFlagsArray`, or we'll have to specify globbed paths inside `preCheck`. Other attributes also enhance the structural selection of the test cases.

This enhancement is essential to the tree-wide transition to deprecate the non `__structuredAttrs`-compatible `pytestFlagsArray` flags. See the previous discussion: https://github.com/NixOS/nixpkgs/pull/347194#issuecomment-2583298737

The keyword and markers expressions shares the same form `((inc1) or (inc2)) and not (exc1) and not (exc2)`, and are passed to `-k` and `-m` respectively.

I have considered renaming them to `includeTests` and `excludeTests`, but I'm unsure if tree-wide renaming from `disabledTests` to `excludeTests` would be worthwhile.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
python3Packages.pytestCheckHook: support deselecting tests via disabledTestPaths

Pass disabledTestPaths elements containing double colons (::)
to --deselect instead of --ignore-glob.

ShamrockLee committed:
python3Packages.pytestCheckHook: support enabledTestPaths

ShamrockLee committed:
python3Packages.pytestCheckHook: append disableTests flags after disabledTestPaths flags

ShamrockLee committed:
python3Packages.pytestCheckHook: support enabledTests

ShamrockLee committed:
python3Packages.pytestCheckHook: support enabledTestMarks and disabledTestMarks

ShamrockLee committed:
python3Packages.xyzservices: use disabledTestMarks

ShamrockLee commented:
Do you think we should rename from `disabledTests` to `excludeTests` and `disabledTestPaths` to `excludeTestPaths`?

This way, we can call the whole set of arguments:
- `includeTestPaths`, `excludeTestPaths`
- `includeTests`, `excludeTests`
- `includeTestMarks`, `excludeTestMarks`

Ping @wolfgangwalther

ShamrockLee committed:
python3Packages.pytestCheckHook: Make disabledTests support sub-expressions

Wrap each elements of disabledTests with parenthesis
so that when __structuredAttrs = true,
people could use sub-expressions an element.

E.g.
```nix
{
  disabledTests = [
    "ClassFoo and test_foo"
    "test_bar"
  ];
}

ShamrockLee committed:
doc: python.section.md: partial rewrites for pytest and pytestCheckHook

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>
Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>
Co-authored-by: Wolfgang Walther <walther@technowledgy.de>

mweinelt commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,54 +1250,39 @@ when a package may need many items disabled to run the test suite.
-for example:
+`pytestCheckHook` recognizes the following attributes/variables to include/whitelist tests and to exclude/blacklist tests:
+
+- `enabledTestPaths` and `disabledTestPath`: to specify path globs (files or directories) or test items.
```
```suggestion
- `enabledTestPaths` and `disabledTestPaths`: to specify path globs (files or directories) or test items.
```

mweinelt reviewed (commented):
Cool.

ShamrockLee commented:
I added a short paragraph about `pytestFlags`.

ShamrockLee commented:
Just fixed the implementation of the newly-added sub-expression feature. `python3Packages.pytestCheckHook.tests` should now build.

ShamrockLee commented:
@ofborf build pythonPackages.pytestCheckHook.tests

ShamrockLee commented:
@ofborf build pythonPackages.xyzservices

SuperSandro2000 commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -439,13 +443,29 @@ let
+          (
+            name: value:
+            lib.throwIf (finalAttrs.doInstallCheck && attrs.${name} == [ ])
+              "${lib.getName finalAttrs}: ${name} must be unspecified, null or a non-empty list when doInstallCheck is true."
```
It should also be that when doInstallCheck is false

SuperSandro2000 commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,54 +1250,43 @@ when a package may need many items disabled to run the test suite.
 so you will most likely use `pytestCheckHook`.
 
-Using the example above, the analogous `pytestCheckHook` usage would be:
+To use `pytstCheckHook`, add it into the `nativeBuildInputs`. (You don't have to add `pytest` explicitly for `pytestCheckHook` to work.)
```
```suggestion
To use `pytstCheckHook`, add it to `nativeBuildInputs`. Adding `pytest` explicitly is not required.
```

ShamrockLee replied:
> Should be `nativeCheckInputs` instead of `nativeBuildInputs`, right?

You're right.

SuperSandro2000 commented:
> I have considered renaming them to `includeTests` and `excludeTests`, but I'm unsure if tree-wide renaming from `disabledTests` to `excludeTests` would be worthwhile.

same vibe.

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,54 +1250,43 @@ when a package may need many items disabled to run the test suite.
 
-This is especially useful when tests need to be conditionally disabled,
-for example:
+`pytestCheckHook` recognizes the following attributes/variables to include/whitelist tests and to exclude/blacklist tests:
```
Simplified wording:
```suggestion
`pytestCheckHook` recognizes the following attributes for whitelisting and/or blacklisting tests:
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,54 +1250,43 @@ when a package may need many items disabled to run the test suite.
+- `enabledTestPaths` and `disabledTestPaths`: to specify path globs (files or directories) or test items.
+- `enabledTests` and `disabledTests`: to specify keywords for class names or test method names.
+- `enabledTestMarks` and `disabledTestMarks`: to specify test marks.
+
```
```suggestion

  "enabled" attributes define a whitelist of tests to run, while "disabled" attributes define a blacklist of tests that will not be run.
  If the whitelist is empty, it is ignored and only the blacklist is considered.
  If a test matches both the whitelist and blacklist, it is not run.

```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,54 +1250,43 @@ when a package may need many items disabled to run the test suite.
-  ] ++ lib.optionals stdenv.buildPlatform.isDarwin [
-    # can fail when building with other packages
-    "socket"
+    "Foo and bar"
```
Again, why remove real world examples?

ShamrockLee replied:
I added it back.

ShamrockLee commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,54 +1250,43 @@ when a package may need many items disabled to run the test suite.
-
-  disabledTestPaths = [
-    "tests/test_failing.py"
-  ];
```
The original example emphasize the transition from using `pytest` command inside `checkPhase` (`installCheckPhase`) to `pytestCheckHook` and is coupled to the historical implementation of the hook. Let me add a new one exclusively for `pytestCheckHook`.

ShamrockLee replied:
Addressed.

ShamrockLee commented:
I removed the duplicated documentation about `pytsetCheckHook` test selection under the `Common issues` heading and merge them to the `pytestCheckHook` subsection.

MattSturgeon reviewed (approved):
No issues from me

ShamrockLee commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -439,13 +443,29 @@ let
+        lib.mapAttrs
+          (
+            name: value:
+            lib.throwIf (attrs.${name} == [ ])
```
Such verbose constraint has an additional impact of enforcing list-based specifications (instead of space-separated strings or empty strings) when `__structuredAttrs = false`, but I haven't seen it elsewhere and I'm unsure if we should do that.

@SuperSandro2000 What do you think?

ShamrockLee commented:
I fixed a typo in the documentation change: `"requires"` -> `"request"`.

wolfgangwalther reviewed (commented):
The terms "whitelist" and "blacklist" don't actually help the understanding, in fact they make it harder to read. I think the simpler language reads better - and is more inclusive as well.

ShamrockLee commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,115 @@ when a package may need many items disabled to run the test suite.
 
-  disabledTestPaths = [
-    "tests/test_failing.py"
+  # Blacklist tests by test marks.
```
I haven'found a real-world example, but specifying `enabledTestMarks` is likely to exclude most of the tests otherwise included.

Still, it is occationally practical to include only a small part of the tests. I just cannot come up with a value that goes well with other attributes values specifed here.

ShamrockLee replied:
> I guess everybody has understood the concept by here, but it's inconsistent. We could probably skip some others, too.

How about we remove `disabledTestMarks` from  the example? The implementation logic is similar to `enabledTests` and `disabledTests`, and maybe `enabledTests` and `disabledTests` are enough to show how to use them.

ShamrockLee replied:
> > How about we remove `disabledTestMarks` from the example?
>
> Fine by me!

Done.

ShamrockLee commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,115 @@ when a package may need many items disabled to run the test suite.
+- `enabledTests` and `disabledTests`: to specify keywords for class names or test method names.
+- `enabledTestMarks` and `disabledTestMarks`: to specify test marks.
+
+"enabled" attributes define a whitelist of tests to run, while "disabled" attributes define a blacklist of tests that will not be run.
+If the whitelist is empty, it is ignored and only the blacklist is considered.
+If a test matches both the whitelist and blacklist, it is not run.
```
That sounds a lot clearer!

Nevertheless, "all tests" here might refer to "all the tests included during pytest's test discovery." If not further configured, it requires matching pytest's default test naming conventions:

```toml
[tool.pytest.ini_options]
python_files = [
    "test_*.py",
    "*_test.py",
]
python_classes = [
    "Test*",
]
pthon_functions = [
    "test*",
]
```

where `python_functions` governs module-level functions or class methods.

The `enable*` attributes are able to add tests outside such naming conventions.

ShamrockLee commented:
The pytest documentation seems to mix the words "marks" and "markers." Here I'll follow the "Reference" documentation and use "marks."

- https://docs.pytest.org/en/stable/reference/reference.html#marks-ref
- https://docs.pytest.org/en/stable/how-to/mark.html#mark
- https://docs.pytest.org/en/stable/example/markers.html

ShamrockLee commented:
I improved the example based on pytest's default test discovery documented in its reference manual.

See the discussion comment for detail: https://github.com/NixOS/nixpkgs/pull/386513#discussion_r1996437109

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
 so you will most likely use `pytestCheckHook`.
 
-Using the example above, the analogous `pytestCheckHook` usage would be:
+To use `pytestCheckHook`, add it to `nativeCheckInputs`. Adding `pytest` explicitly is not required.
```
```suggestion
To use `pytestCheckHook`, add it to `nativeCheckInputs`.
Adding `pytest` is not required, since it is included with `pytestCheckHook`.
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
````diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
+}
+```
+
+`pytestCheckHook` recognizes the following attributes to include or exclude tests:
````
We could probably simplify this down to:
```suggestion
`pytestCheckHook` recognizes the following attributes:
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
+
+`pytestCheckHook` recognizes the following attributes to include or exclude tests:
+
+- `enabledTestPaths` and `disabledTestPaths`: to specify path globs (files or directories) or test items.
+- `enabledTests` and `disabledTests`: to specify keywords for class names or test method names.
+- `enabledTestMarks` and `disabledTestMarks`: to specify test marks.
```
Should we be using definition-list syntax here?

```suggestion
`enabledTestPaths` and `disabledTestPaths`

:   To specify path globs (files or directories) or test items.

`enabledTests` and `disabledTests`

:   To specify keywords for class names or test method names.

`enabledTestMarks` and `disabledTestMarks`

:   To specify test marks.
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
+If tests are explicitly enabled, only those tests will run.
+A test, that is both enabled and disabled, will not run.
+
+The following example demonstrates the use of the "enabled" and "disabled" pytestCheckHook attributes:
```
```suggestion
The following example demonstrates usage of various `pytestCheckHook` attributes:
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
 
-  disabledTestPaths = [
-    "tests/test_failing.py"
+The `<enabled/disabled>TestsPaths` expands Unix-style globs.
```
```suggestion
The `<enabled/disabled>TestsPaths` attributes expand Unix-style globs.
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
-  disabledTestPaths = [
-    "tests/test_failing.py"
+The `<enabled/disabled>TestsPaths` expands Unix-style globs.
+If the test paths to include/exclude contain characters like `*`, `?`, `[`, and `]`,
```
```suggestion
If a test path contains characters like `*`, `?`, `[`, or `]`, you can
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
```diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
+which will be passed to pytest's `-k` and `-m` flags respectively.
+With `__structuredAttrs = true` enabled, they additionally support sub-expressions.
+
+For example,
+the following `disabledTests` specification excludes test items like
+`TestFoo::test_bar_functionality`:
```
```suggestion
For example, you could disable test items like `TestFoo::test_bar_functionality`
by disabling tests that match both `"Foo"` **and** `"bar"`:
```

MattSturgeon commented on doc/languages-frameworks/python.section.md:
````diff
@@ -1272,34 +1250,105 @@ when a package may need many items disabled to run the test suite.
+}
+```
+
+The main benefits of constructing `pytest` commands from `pytestCheckHook`-supported attributes/variables
````
```suggestion
The main benefits of using `pytestCheckHook` to construct `pytest` commands
```

MattSturgeon reviewed (approved):
A few minor suggestions, nothing blocking

ShamrockLee commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -455,7 +455,7 @@ let
         lib.mapAttrs
           (
             name: value:
-            lib.throwIf (finalAttrs.doInstallCheck && attrs.${name} == [ ])
```
I didn't notice that when committing [Sandro's suggestion of removing the `finalAttrs.doInstallCheck` condition from the throwing](https://github.com/NixOS/nixpkgs/pull/386513#discussion_r1991352139). Sorry for the confusion.

> My naive understanding is that you just want to throw an error when the user makes a mistake and:
> 
>     * enables the check phase
> 
>     * does not enable any tests

Exactly.

> If that's true, then why remove the install check condition?

The implication behind the choice to consider `finalAttrs.doInstallCheck` in the assertion would be tolerating the existence of `enabledTests = [ ]` in addition to `enabledTests = null`. That would increase the cases to consider in derivation overriding.

ShamrockLee replied:
Addressed.

ShamrockLee commented on pkgs/development/interpreters/python/hooks/pytest-check-hook.sh:
```diff
@@ -10,7 +10,33 @@ function pytestCheckPhase() {
+        if [[ "$path" =~ "::" ]]; then
+            flagsArray+=("$path")
+        else
+            # The `|| kill "$$"` trick propagates the errors from the process substitutiton subshell,
+            # which is suggested by a StackOverflow answer: https://unix.stackexchange.com/a/217643
+            readarray -t -O"${#flagsArray[@]}" flagsArray < <(@pythonCheckInterpreter@ - "$path" <<EOF || kill "$$")
```
The Bash expressions after the pipe is executed in a sub-shell and the `flagsArray` change there won't come out.

ShamrockLee commented:
@MattSturgeon, I addressed the suggestions and adopted [the definition-list syntax](https://github.com/NixOS/nixpkgs/pull/386513#discussion_r1997665205).

I also add `pytestFlags` attribute description to the list, and merge its `pytestFlags` example into the main example. I moved the `__structuredAttrs` part to a new paragraph below the main example.

wolfgangwalther reviewed (approved):
Did not do any builds or tests, but code LGTM.

(short of the current CI failure for nixfmt)

ShamrockLee commented:
The current assertion only guard the `buildPythonPackage` (and `overridePythonAttrs`) arguments but not the overridden attributes by `overrideAttrs`, since the latter is tricky to implement.

ShamrockLee commented:
> Did not do any builds or tests, but code LGTM.

Let's see if `python3Packages.pytestCheckHook.tests` and `python3Packages.xyzservices` build.

@ofborg build nixpkgs-manual

ShamrockLee commented:
> (short of the current CI failure for nixfmt)

Addressed.

ShamrockLee commented:
The OfBorg CI test building xyzservices on aarch64-linux fails because an inline test of `perl5` doesn't pass. This is unrelated to this PR, but might hint that it's a bit unrealistic to expect `python3Packages.xyzservices` to build smoothly on the staging branch.

The tests should be lighter:
@ofborg build python3Packages.pytestCheckHook.tests

ShamrockLee commented:
I cherry-picked the documentation changes onto nixos-unstable-small and the Nixpkgs Manual builds fine.

OfBorg's failures to build `nixpkgs-manual` on aarch64-linux seem to caused by a Perl plugin failing to pass three of its tests. That's not related to the issue.

Partial log:

```
error: builder for '/nix/store/bpl8lqqw67kpqd7jcqy0km6fn3rs67i9-perl5.40.0-XML-Parser-2.46.drv' failed with exit code 2;
       last 25 log lines:
       > PERL_DL_NONLAZY=1 "/nix/store/ds9x5fnzhspvcndr32i9qf5d95a3xd57-perl-5.40.0/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
       > t/astress.t ........... ok
       > t/cdata.t ............. ok
       > t/decl.t .............. ok
       > t/defaulted.t ......... ok
       > t/encoding.t .......... ok
       > t/external_ent.t ...... ok
       > t/file.t .............. ok
       > t/file_open_scalar.t .. ok
       > t/finish.t ............ ok
       > t/namespaces.t ........ ok
       > t/parament.t .......... ok
       > t/partial.t ........... Failed 1/3 subtests
       > t/skip.t .............. ok
       > t/stream.t ............ ok
       > t/styles.t ............ ok
       >
       > Test Summary Report
       > -------------------
       > t/partial.t         (Wstat: 0 Tests: 3 Failed: 1)
       >   Failed test:  3
       > Files=15, Tests=140,  2 wallclock secs ( 0.06 usr  0.03 sys +  0.73 cusr  0.49 csys =  1.31 CPU)
       > Result: FAIL
       > Failed 1/15 test programs. 1/140 subtests failed.
       > make: *** [Makefile:981: test_dynamic] Error 255
```

That is unrelated to this issue.

Given that `python3Packages.pytestCheckHook.tests` builds on aarch64-linux, and that the functionality is platform-agnostic, I'm going to merge this PR.

ShamrockLee commented on pkgs/development/interpreters/python/hooks/pytest-check-hook.sh:
```diff
@@ -3,22 +3,64 @@
+        else
+            # The `|| kill "$$"` trick propagates the errors from the process substitutiton subshell,
+            # which is suggested by a StackOverflow answer: https://unix.stackexchange.com/a/217643
+            readarray -t -O"${#flagsArray[@]}" flagsArray < <(@pythonCheckInterpreter@ - "$path" <<EOF || kill "$$")
```
> I'm afraid this doesn't work as expected,
> 
> In a [random hydra log](https://hydra.nixos.org/build/296020849/nixlog/1), I found this error:
> 
> ```
> /nix/store/l3797z8ybpgk14ara7z0ld4hsnrpnvyv-pytest-check-hook/nix-support/setup-hook: line 41: warning: command substitution: 1 unterminated here-document
> ```
> 
> This makes sense - the here doc starts in the subshell, but ends outside.
> 

I'll fix it this weekend.

> What I am confused about even more is the fact that this doesn't break the build at this stage, yet.

I guess it's because the syntax error require a missing EOF for the heredoc, which is only valid when taking account the whole line. Since it the doesn't happen before the `||` operator, it isn't caught by the kill command.
